# Scipy

In [ ]:
pip install scipy

Version

In [ ]:
import scipy as sp

print("SciPy version:{}".format(sp.__version__))

## Find Cubic Root

In [1]:
from scipy.special import cbrt

#Find cubic root of 27 & 64 using cbrt() function
cubic_root = cbrt([27, 64])

In [2]:
#Print content of cubic_root
print("cubic roots: {}".format(cubic_root))

cubic roots: [3. 4.]


## Exponential Values

In [ ]:
from scipy.special import exp10

#Define exp10 function and pass value in its
exp_values = exp10([1,10])

In [ ]:
#Print content of exp_values
print("exponential values: {}".format(exp_values))

## Combinantion

The number of combinations of N things taken k at a time.

Funzione che calcola il numero di combinazioni (o sottoinsiemi) possibili di un insieme di N elementi, presi k alla volta. 

In [5]:
from scipy.special import comb

#Find combinations of 5, 2 values using comb(N, k)
com = comb(N=5, k=2, exact = False, repetition=True)

La funzione calcola il numero di combinazioni di 2 elementi scelti da un insieme di 5 elementi, con ripetizione.

La formula per calcolare le combinazioni con ripetizione è: 

$ \binom{N+k-1}{k} $


In [6]:
#Print content of com
print("combination value: {}".format(com))

combination value: 15.0


## Matrix determinant

In [ ]:
from scipy import linalg
import numpy as np

#Define 2d NumPy array
arr_2d = np.array([ (4,5), (3,2) ])

In [ ]:
#Print content of arr_2
print ("arr_2d content: {}".format(arr_2d))

In [ ]:
#Pass values to det() function
det = linalg.det(arr_2d)

In [ ]:
#Print content of det
print ("matrix determinant: {}".format(det))

## Inverse matrix

In [ ]:
from scipy import linalg
import numpy as np

#Define 2d NumPy array
arr_2d = np.array([ [4,5], [3,2] ])

In [ ]:
#Print content of arr_2d
print ("arr_2d content: {}".format(arr_2d))

In [ ]:
#Pass value to function inv()
inv = linalg.inv( arr_2d )

In [ ]:
#Print content of inv
print ("inverse matrix : {}".format(inv))

## Autovalori vs Autovettori

In [ ]:
from scipy import linalg
import numpy as np

#Define 2d NumPy array
arr_2d = np.array([[5,4],[6,3]])

In [ ]:
#Pass value into function eig()
eg_val, eg_vect = linalg.eig(arr_2d)

In [ ]:
#Print content of eg_val
print("eigenvalues: {}".format(eg_val))

In [ ]:
#Print content of eg_vect
print("eigenvectors: {}".format(eg_vect))

# Integrazione Numerica Parte 1

il modulo integrate dalla libreria scipy, che fornisce funzioni per eseguire integrazioni numeriche. Una delle funzioni principali di questo modulo è quad, che permette di calcolare l'integrale definito di una funzione.

Definizione della funzione da integrare

In [10]:
from scipy import integrate

# Take f(x) function as f
f = lambda x : x**2

stiamo definendo una funzione usando una lambda function (funzione anonima). La funzione è molto semplice: restituisce x^2, quindi  
𝑓(𝑥)=x^2.

La funzione integrate.quad viene utilizzata per calcolare l'integrale definito di una funzione su un intervallo. 

In [12]:
#Single integration with a = 0 & b = 1
integration = integrate.quad(f, 0 , 1)

La variabile integration è una tupla, quindi il risultato che verrà stampato avrà il seguente formato:

Integrazione analitica:

Se calcoliamo l'integrale manualmente:

$
\int_0^1 x^2 \, dx = \left[ \frac{x^3}{3} \right]_0^1 = \frac{1^3}{3} - \frac{0^3}{3} = \frac{1}{3}
$

In [13]:
#Print content of integration
print("integration:{}".format(integration))

integration:(0.33333333333333337, 3.700743415417189e-15)


- 0.33333333333333337 è il valore calcolato numericamente dell'integrale.
- 3.700743415417189e-11 è una stima dell'errore numerico associato al calcolo.

# Integrazione Numerica Parte 2

Perform double integration

In [ ]:
from scipy import integrate

#Import square root function from math lib
from math import sqrt

In [ ]:
# set  function f(x)
f = lambda x, y : 64 *x*y

# Lower limit of second integral
p = lambda x : 0

# Upper limit of first integral
q = lambda y : sqrt(1 - 2*y**2)

In [ ]:
# Perform double integration
integration = integrate.dblquad(f , 0 , 2/4,  p, q)

In [ ]:
#Print content of integration
print("integration: {}".format(integration))